server/server.go

```diff
@@ -0,0 +1,301 @@
+package server
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	iavl "github.com/cosmos/iavl"
+	pb "github.com/cosmos/iavl/proto"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var _ pb.IAVLServiceServer = (*IAVLServer)(nil)
+
+// IAVLServer implements the gRPC IAVLServiceServer interface. It provides a gRPC
+// API over an IAVL tree.
+// rwLock is used to ensure:
+// 1. No reading while writing, no writing while reading.
+// 2. Unlimited concurrent reads.
+// 3. Sequential writes.
+type IAVLServer struct {
```

Comment:
The server does not appear to have any panic handlers installed, such that a panic causes the entire server to crash. This is particularly problematic since IAVL uses panics as a primary error mechanism. We should use some sort of middleware which catches the panics and converts them into error responses.

Reply:
copy-pasted the solution from http://www.inanzzz.com/index.php/post/oicw/panic-recovery-middleware-for-golang-http-server